library/sets/set1.lgt

```diff
@@ -19,17 +19,81 @@
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 
 
-:- object(set(_Type),
+:- object(set(_Type_),
 	extends(set)).
 
 	:- info([
-		version is 1:23:0,
-		author is 'Paulo Moura',
-		date is 2018-07-11,
-		comment is 'Set predicates with elements constrained to a single type.',
+		version is 1:24:0,
+		author is 'Paulo Moura and Adrian Arroyo',
+		date is 2022-02-03,
+		comment is 'Set predicates with elements constrained to a single type and custom comparing rules.',
 		parnames is ['Type']
 	]).
 
+	:- private(sort/2).
+	:- mode(sort(+list, -list), one).
+	:- info(sort/2, [
+		comment is 'Sorts a list in ascending order.',
+		argnames is ['List', 'Sorted']
+	]).
+
+	:- private(partition/4).
+	:- mode(partition(+list, +nonvar, -list, -list), one).
+	:- info(partition/4, [
+		comment is 'List partition in two sub-lists using a pivot.',
+		argnames is ['List', 'Pivot', 'Lowers', 'Biggers']
+	]).
+
+	:- private(compare/3).
+	:- info(compare/3, [
+		comment is 'Compares two elements using custo comparing rules.',
+		argnames is ['Order', 'Term1', 'Term2']
+	]).
+
+	sort([], []).
+	sort([P| L], S) :-
+		partition(L, P, Small, Large),
+		sort(Small, S0),
+		sort(Large, S1),
+		list::append(S0, [P| S1], S).
+
+	partition([], _, [], []).
+	partition([X| L1], P, Small, Large) :-
+		(	_Type_::(X < P) ->
+			Small = [X| Small1], Large = Large1
+		;	_Type_::(X =:= P) ->
+            Small = Small1, Large = Large1  
+        ;   Small = Small1, Large = [X| Large1]
+		),
+		partition(L1, P, Small1, Large1).
+
+	compare(<, Term1, Term2) :-
+		_Type_::(Term1 < Term2),!.
+
+	compare(=, Term1, Term2) :-
+		_Type_::(Term1 =:= Term2),!.
+
+	compare(>, Term1, Term2) :-
+		_Type_::(Term1 > Term2),!.
+
+	as_set(List, Set) :-
+		sort(List, Set).
+
+	insert([], Element, [Element]).
+	insert([Head| Tail], Element, Set) :-
+		compare(Order, Head, Element),
```

Comment:
We can improve performance by using an if-then-else in the definition of the `insert/3` predicate without using the custom `compare/3` predicate, which implies creating a choice-point per call followed by its destruction when reaching the cut.

Reply:
Good idea!